### PR TITLE
Import the snapshot header

### PR DESCRIPTION
**Pull request by Remagpie**
Depends: https://github.com/CodeChain-io/foundry/pull/19
Collected from https://github.com/CodeChain-io/codechain/pull/1855 and https://github.com/CodeChain-io/codechain/pull/1862

### CODE REVIEW

File: sync/src/block/extension.rs [hunk 1]
```diff
@@ -103,6 +118,45 @@ impl Extension {
         }
     }
 
+    fn initial_state(client: Arc<Client>, snapshot_target: Option<(H256, u64)>) -> State {
```
Comment (sgkim126) [inline sync/src/block/extension.rs]: How about making it a constructor of `State`?

Comment (Remagpie) [reply sync/src/block/extension.rs]: Actually there's a patch that moves this function to `State::initial` method. I'll move that patch to here.

File: util/merkle/src/triedbmut.rs [hunk 2]
```diff
@@ -417,6 +417,10 @@ impl<'a> Trie for TrieDBMut<'a> {
 
         t.get(key)
     }
+
+    fn is_complete(&self) -> bool {
+        TrieDB::try_new(self.db, self.root).map(|t| t.is_complete()).unwrap_or(false)
```
Comment (sgkim126) [inline util/merkle/src/triedbmut.rs]:
```rust
TrieDB::try_new(self.db, self.root).map_or(false, |t| t.is_complete())
```

Comment (Remagpie) [reply util/merkle/src/triedbmut.rs]: [`Result`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) doesn't have `map_or` function.

Comment (sgkim126) [reply util/merkle/src/triedbmut.rs]: https://doc.rust-lang.org/nightly/std/result/enum.Result.html#method.map_or Sorry. It is nightly yet.

File: foundry/foundry.yml [hunk 3]
```diff
@@ -252,6 +252,16 @@ args:
         takes_value: true
         conflicts_with:
             - no-discovery
+    - snapshot-hash:
```
Comment (sgkim126) [inline foundry/foundry.yml]: Would you describe what this option does?

File: foundry/foundry.yml [hunk 4]
```diff
@@ -252,6 +252,16 @@ args:
...
+        long: snapshot-hash
+        value_name: HASH
+        requires: snapshot-number
+        takes_value: true
+    - snapshot-number:
```
Comment (sgkim126) [inline foundry/foundry.yml]: ditto.

File: foundry/foundry.yml (hunk 3 above)
Comment (Remagpie) [reply foundry/foundry.yml]: Added help message to this option.

File: foundry/foundry.yml (hunk 4 above)
Comment (Remagpie) [reply foundry/foundry.yml]: Added a help message for this option.

**Pull request merged**